File: 2018/day04_2.py
# pylint: disable=missing-module-docstring, missing-function-docstring


import re
from datetime import datetime, timedelta
from typing import Any

type Guard = tuple[int, str, str]


def solution(content: list[str]) -> Any:
    guards = group_guard_and_minute(parse(sorted(content)))
    _, key, val = sorted(
        [(v, m, key) for key, guard in guards.items() for m, v in guard.items()],
        reverse=True,
    )[0]
    return key * val


def group_guard_and_minute(timeshift: list[Guard]):
    guards: dict[int, dict[int, int]] = {}
    for shift in timeshift:
        guard_id, start, end = shift
        guards[guard_id] = guards.get(guard_id, {})
        current = datetime.fromisoformat(start)
        end_dt = datetime.fromisoformat(end)
        while current < end_dt:
            guards[guard_id][current.minute] = (
                guards[guard_id].get(current.minute, 0) + 1
            )
            current = current + timedelta(minutes=1)
    return guards


def parse(lines: list[str]):
    guard_id = None
    output = []
    start = None
    for line in lines:
        matches = re.match(r"\[(\d{4}-\d{2}-\d{2} \d{2}:\d{2})\] (.+)", line)
        if matches is None:
            raise ValueError(f"Invalid line {line}")
        begins = re.match(r"Guard #(\d+) begins shift", matches.group(2))
        asleep = "falls asleep" == matches.group(2)
        wakesup = "wakes up" == matches.group(2)
        if begins is not None:
            guard_id = begins.group(1)
        elif asleep:
            start = matches.group(1)
        elif wakesup and guard_id is not None:
            output.append((int(guard_id), start, matches.group(1)))
    return output


def main() -> None:
    with open("day04.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()

File: 2018/day04-2.ts
type Timeshift = {
  id: string;
  start: Date;
  end: Date;
};

function solution(content: string) {
  const [key, [val]] = mostAsleepGuard(
    mostAsleepMinute(
      groupByGuardAndMinute(parse(content.split("\n").toSorted()))
    )
  );
  return Number(key) * Number(val);
}

function mostAsleepGuard(guards: Record<string, [string, number]>) {
  return Object.entries(guards).toSorted(
    ([_aKey, [_aMin, aVal]], [_bKey, [_bMin, bVal]]) => bVal - aVal
  )[0];
}

function mostAsleepMinute(guards: Record<string, Record<number, number>>) {
  return Object.fromEntries(
    Object.entries(guards).map(([id, guard]) => [
      id,
      Object.entries(guard).toSorted(
        ([_aId, aVal], [_bId, bVal]) => bVal - aVal
      )[0],
    ])
  );
}

function groupByGuardAndMinute(timeshifts: Timeshift[]) {
  const output: Record<string, Record<number, number>> = {};

  timeshifts.forEach(({ id, end, start }) => {
    if (output[id] === undefined) {
      output[id] = {};
    }
    const current = new Date(start);
    while (current < end) {
      const minute = current.getMinutes();
      output[id][minute] = (output[id][minute] ?? 0) + 1;
      current.setMinutes(minute + 1);
    }
  });

  return output;
}

function parse(lines: string[]): Timeshift[] {
  let guardId: string | null = null;
  const output: Timeshift[] = [];
  let start: Date | null = null;
  lines.forEach((line) => {
    const matches = line.match(/\[(\d{4}-\d{2}-\d{2} \d{2}:\d{2})\] (.+)/);
    if (matches === null) throw new Error(`Invalid line: ${line}`);
    const date = matches.at(1);
    const note = matches.at(2);
    if (!date || !note) throw new Error(`Invalid line: ${line}`);

    const begins = note.match(/Guard #(\d+) begins shift/);
    const asleep = note.includes("falls asleep");
    const wakeup = note.includes("wakes up");

    if (begins !== null) {
      guardId = begins.at(1) ?? "";
    } else if (asleep) {
      start = new Date(date);
    } else if (wakeup && guardId !== null && start !== null) {
      output.push({
        id: guardId,
        start,
        end: new Date(date),
      });
    }
  });

  return output;
}

Deno.readTextFile("day04.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
